File: working_hours/domain/parser.py
import re
import sys
from datetime import datetime
from pathlib import Path

from .models import ClockEvent, CorrectionItem


def parse_file(path: Path) -> list[ClockEvent]:
    """Parse a tab-separated clocking export file.

    Skips header lines and lines starting with # (comment/correction markers).
    Normalises inconsistent whitespace in the timestamp column.
    """
    events: list[ClockEvent] = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.rstrip("\n")
            if line.startswith("#"):
                continue
            parts = line.split("\t")
            if len(parts) < 4:
                continue
            raw_id = parts[0].strip().rstrip(".")
            if not raw_id.isdigit():
                continue
            name = parts[1].strip()
            dept = parts[2].strip()
            time_str = re.sub(r"\s+", " ", parts[3].strip())
            try:
                ts = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
            except ValueError:
                continue
            events.append(ClockEvent(raw_id, name, dept, ts))
    return events


def _parse_event_parts(parts: list[str]) -> ClockEvent | None:
    """Parse [id, name, dept, timestamp, ...] into a ClockEvent."""
    if len(parts) < 4:
        return None
    raw_id = parts[0].strip().rstrip(".")
    if not raw_id.isdigit():
        return None
    name = parts[1].strip()
    dept = parts[2].strip()
    time_str = re.sub(r"\s+", " ", parts[3].strip())
    try:
        ts = datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
    except ValueError:
        return None
    return ClockEvent(raw_id, name, dept, ts)


def parse_correction_file(path: Path) -> list[CorrectionItem]:
    """Parse a corrections file with optional ADD/DEL/EDIT action prefixes.

    Plain lines (no prefix) are treated as ADD for backwards compatibility.
    EDIT format: EDIT <tab> id <tab> name <tab> dept <tab> old_ts <tab> new_ts [<tab> device]
    """
    items: list[CorrectionItem] = []
    with open(path, encoding="utf-8") as f:
        for line in f:
            line = line.rstrip("\n")
            if line.startswith("#") or not line.strip():
                continue
            parts = line.split("\t")
            first = parts[0].strip()

            if first in ("ADD", "DEL", "EDIT"):
                action = first
                rest = parts[1:]
            else:
                action = "ADD"
                rest = parts

            if action == "EDIT":
                # EDIT: id, name, dept, old_ts, new_ts[, device_id]
                if len(rest) < 5:
                    continue
                raw_id = rest[0].strip().rstrip(".")
                if not raw_id.isdigit():
                    continue
                name = rest[1].strip()
                dept = rest[2].strip()
                old_ts_str = re.sub(r"\s+", " ", rest[3].strip())
                new_ts_str = re.sub(r"\s+", " ", rest[4].strip())
                try:
                    old_ts = datetime.strptime(old_ts_str, "%Y-%m-%d %H:%M:%S")
                    new_ts = datetime.strptime(new_ts_str, "%Y-%m-%d %H:%M:%S")
                except ValueError:
                    continue
                items.append(CorrectionItem(
                    action="EDIT",
                    event=ClockEvent(raw_id, name, dept, new_ts),
                    old_timestamp=old_ts,
                ))
            else:
                event = _parse_event_parts(rest)
                if event is None:
                    continue
                items.append(CorrectionItem(action=action, event=event, old_timestamp=None))
    return items


def apply_corrections(raw: list[ClockEvent], corrections: list[CorrectionItem]) -> list[ClockEvent]:
    """Apply ADD/DEL/EDIT corrections to a raw event list, in order."""
    events = list(raw)
    for item in corrections:
        if item.action == "ADD":
            events.append(item.event)
        elif item.action == "DEL":
            events = [
                e for e in events
                if not (e.emp_id == item.event.emp_id and e.timestamp == item.event.timestamp)
            ]
        elif item.action == "EDIT":
            events = [
                ClockEvent(e.emp_id, e.name, e.dept, item.event.timestamp)
                if (e.emp_id == item.event.emp_id and e.timestamp == item.old_timestamp)
                else e
                for e in events
            ]
    return events


def validate_raw_content(content: str) -> tuple[bool, str]:
    """Return (True, '') if content has at least one parseable clock event."""
    for line in content.splitlines():
        if line.startswith("#"):
            continue
        parts = line.split("\t")
        if len(parts) < 4:
            continue
        raw_id = parts[0].strip().rstrip(".")
        if not raw_id.isdigit():
            continue
        time_str = re.sub(r"\s+", " ", parts[3].strip())
        try:
            datetime.strptime(time_str, "%Y-%m-%d %H:%M:%S")
            return True, ""
        except ValueError:
            continue
    return False, "No valid clock events found. Expected tab-separated rows: ID, Name, Dept, Timestamp (YYYY-MM-DD HH:MM:SS)."


def collect_raw_files(args: list[str]) -> list[Path]:
    """Resolve CLI arguments to raw input .txt files.

    With no arguments, reads input_data/ automatically.
    """
    if not args:
        d = Path("input_data")
        return sorted(d.glob("*.txt")) if d.is_dir() else []

    paths: list[Path] = []
    for arg in args:
        p = Path(arg)
        if p.is_dir():
            paths.extend(sorted(p.glob("*.txt")))
        elif p.is_file():
            paths.append(p)
        else:
            print(f"Warning: {arg!r} not found, skipping.", file=sys.stderr)
    return paths


def collect_correction_files() -> list[Path]:
    """Return all .txt files from the corrections/ directory."""
    d = Path("corrections")
    return sorted(d.glob("*.txt")) if d.is_dir() else []


def collect_files(args: list[str]) -> list[Path]:
    """Resolve CLI arguments to a list of .txt files to process.

    With no arguments, reads input_data/ and corrections/ automatically.
    """
    if not args:
        paths: list[Path] = []
        for d in (Path("input_data"), Path("corrections")):
            if d.is_dir():
                paths.extend(sorted(d.glob("*.txt")))
        return paths

    paths = []
    for arg in args:
        p = Path(arg)
        if p.is_dir():
            paths.extend(sorted(p.glob("*.txt")))
        elif p.is_file():
            paths.append(p)
        else:
            print(f"Warning: {arg!r} not found, skipping.", file=sys.stderr)
    return paths

File: packages/server/src/domain/parser.ts
import { readFileSync } from 'fs'
import type { ClockEvent, CorrectionItem } from './models.js'

const TS_FORMAT = /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$/

function parseTimestamp(raw: string): Date | null {
  const s = raw.replace(/\s+/g, ' ').trim()
  if (!TS_FORMAT.test(s)) return null
  const d = new Date(s.replace(' ', 'T'))
  return isNaN(d.getTime()) ? null : d
}

function parseEventParts(parts: string[]): ClockEvent | null {
  if (parts.length < 4) return null
  const rawId = parts[0].trim().replace(/\.$/, '')
  if (!/^\d+$/.test(rawId)) return null
  const name = parts[1].trim()
  const dept = parts[2].trim()
  const ts = parseTimestamp(parts[3])
  if (!ts) return null
  return { empId: rawId, name, dept, timestamp: ts }
}

export function parseFile(path: string): ClockEvent[] {
  const events: ClockEvent[] = []
  const lines = readFileSync(path, 'utf-8').split('\n')
  for (const raw of lines) {
    const line = raw.replace(/\r$/, '')
    if (line.startsWith('#')) continue
    const parts = line.split('\t')
    const ev = parseEventParts(parts)
    if (ev) events.push(ev)
  }
  return events
}

export function parseCorrectionFile(path: string): CorrectionItem[] {
  const items: CorrectionItem[] = []
  const lines = readFileSync(path, 'utf-8').split('\n')
  for (const raw of lines) {
    const line = raw.replace(/\r$/, '')
    if (line.startsWith('#') || !line.trim()) continue
    const parts = line.split('\t')
    const first = parts[0].trim()

    let action: 'ADD' | 'DEL' | 'EDIT'
    let rest: string[]

    if (first === 'ADD' || first === 'DEL' || first === 'EDIT') {
      action = first
      rest = parts.slice(1)
    } else {
      action = 'ADD'
      rest = parts
    }

    if (action === 'EDIT') {
      if (rest.length < 5) continue
      const rawId = rest[0].trim().replace(/\.$/, '')
      if (!/^\d+$/.test(rawId)) continue
      const name = rest[1].trim()
      const dept = rest[2].trim()
      const oldTs = parseTimestamp(rest[3])
      const newTs = parseTimestamp(rest[4])
      if (!oldTs || !newTs) continue
      items.push({ action: 'EDIT', event: { empId: rawId, name, dept, timestamp: newTs }, oldTimestamp: oldTs })
    } else {
      const ev = parseEventParts(rest)
      if (!ev) continue
      items.push({ action, event: ev, oldTimestamp: null })
    }
  }
  return items
}

export function applyCorrections(raw: ClockEvent[], corrections: CorrectionItem[]): ClockEvent[] {
  let events = [...raw]
  for (const item of corrections) {
    if (item.action === 'ADD') {
      events.push(item.event)
    } else if (item.action === 'DEL') {
      events = events.filter(
        e => !(e.empId === item.event.empId && e.timestamp.getTime() === item.event.timestamp.getTime())
      )
    } else if (item.action === 'EDIT' && item.oldTimestamp) {
      events = events.map(e =>
        e.empId === item.event.empId && e.timestamp.getTime() === item.oldTimestamp!.getTime()
          ? { ...e, timestamp: item.event.timestamp }
          : e
      )
    }
  }
  return events
}

export function validateRawContent(content: string): { ok: boolean; error: string } {
  for (const raw of content.split('\n')) {
    const line = raw.replace(/\r$/, '')
    if (line.startsWith('#')) continue
    const parts = line.split('\t')
    if (parts.length < 4) continue
    const rawId = parts[0].trim().replace(/\.$/, '')
    if (!/^\d+$/.test(rawId)) continue
    if (parseTimestamp(parts[3])) return { ok: true, error: '' }
  }
  return {
    ok: false,
    error: 'No valid clock events found. Expected tab-separated rows: ID, Name, Dept, Timestamp (YYYY-MM-DD HH:MM:SS).',
  }
}
